### PR TITLE
Migrate ztl sources to ingest

### LINKED ISSUE #1246: Ingest - migrate ztl data sources
- [x] dcp_commercialoverlay
- [x] dcp_limitedheight
- [x] dcp_specialpurpose
- [x] dcp_zoningdistricts
- [x] dof_dtm
- [x] dcp_zoningmapamendments
- [x] dcp_zoningmapindex
- [x] dcp_zoningtaxlots

And we might as well migrate other datasets which come from `edm-publishing/datasets`

- [x] dcp_addresspoinst
- [x] dcp_dcmstreetcenterline
- [x] dcp_edesignation
- [x] dcp_mih
- [x] dcp_specialpurposesubdistricts
- [x] dof_condo
- [x] dof_shoreline

## dcp_commercialoverlay

Non-keyed report
```
________________________________________________________________________________
Row count
    Left: 9634
    Right: 9634
________________________________________________________________________________
Column comparison
    Both
        data_library_version
        ogc_fid
        overlay
        shape_area
        shape_leng
        wkb_geometry
    Left only: None
    Right only: None
    Type differences
        Overlay
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Compared columns
        overlay
        shape_area
        shape_leng
        wkb_geometry
    Left only
        Empty DataFrame
        Columns: [shape_area, shape_leng, overlay, wkb_geometry]
        Index: []
    Right only
        Empty DataFrame
        Columns: [shape_area, shape_leng, overlay, wkb_geometry]
        Index: []
```

## dcp_limitedheight

Non-keyed report
```
________________________________________________________________________________
Tables
    Left: dcp_limitedheight_library
    Right: dcp_limitedheight_ingest
________________________________________________________________________________
Row count
    Left: 14
    Right: 14
________________________________________________________________________________
Column comparison
    Both
        data_library_version
        lhlbl
        lhname
        ogc_fid
        shape_area
        shape_leng
        wkb_geometry
    Left only: None
    Right only: None
    Type differences
        Lhlbl
            Left: character varying
            Right: text
        Lhname
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Compared columns
        lhlbl
        lhname
        shape_area
        shape_leng
        wkb_geometry
    Left only
        Empty DataFrame
        Columns: [shape_area, lhlbl, shape_leng, wkb_geometry, lhname]
        Index: []
    Right only
        Empty DataFrame
        Columns: [shape_area, lhlbl, shape_leng, wkb_geometry, lhname]
        Index: []
```

## dcp_specialpurpose
initially saw differences prompting tweaks in #1236 . After that, things look good:
```
________________________________________________________________________________
Tables
    Left: dcp_specialpurpose_library
    Right: dcp_specialpurpose_ingest
________________________________________________________________________________
Row count
    Left: 93
    Right: 93
________________________________________________________________________________
Column comparison
    Both
        data_library_version
        ogc_fid
        sdlbl
        sdname
        shape_area
        shape_leng
        wkb_geometry
    Left only: None
    Right only: None
    Type differences
        Sdname
            Left: character varying
            Right: text
        Sdlbl
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Key columns
        sdlbl
    Left only: None
    Right only: None
    Columns with diffs: None
    Differences by column: None
```

## dcp_zoningdistricts
Post `make_valid` tweak in #1236, outputs are equivalent
```
________________________________________________________________________________
Tables
    Left: dcp_zoningdistricts_library
    Right: dcp_zoningdistricts_ingest
________________________________________________________________________________
Row count
    Left: 5380
    Right: 5380
________________________________________________________________________________
Column comparison
    Both
        data_library_version
        ogc_fid
        shape_area
        shape_leng
        wkb_geometry
        zonedist
    Left only: None
    Right only: None
    Type differences
        Zonedist
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Compared columns
        shape_area
        shape_leng
        wkb_geometry
        zonedist
    Left only
        Empty DataFrame
        Columns: [shape_area, wkb_geometry, zonedist, shape_leng]
        Index: []
    Right only
        Empty DataFrame
        Columns: [shape_area, wkb_geometry, zonedist, shape_leng]
        Index: []
```

## dcp_zoningmapamendments

currently get error
```
ProgrammingError: (psycopg2.errors.UndefinedFunction) operator does not exist: 
date = text
LINE 7:             WHERE "left"."effective" IS DISTINCT FROM "right...
                                             ^
HINT:  No operator matches the given name and argument types. You might need to 
add explicit type casts.
```

library is coercing the column to date, while ingest it's text

Fixed with dtype coercion preprocessor
```
________________________________________________________________________________
Tables
    Left: dcp_zoningmapamendments_library
    Right: dcp_zoningmapamendments_ingest
________________________________________________________________________________
Row count
    Left: 1333
    Right: 1333
________________________________________________________________________________
Column comparison
    Both
        data_library_version
        effective
        lucats
        ogc_fid
        project_na
        shape_area
        shape_leng
        status
        ulurpno
        wkb_geometry
    Left only: None
    Right only: None
    Type differences
        Lucats
            Left: character varying
            Right: text
        Project na
            Left: character varying
            Right: text
        Ulurpno
            Left: character varying
            Right: text
        Status
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Compared columns
        effective
        lucats
        project_na
        shape_area
        shape_leng
        status
        ulurpno
        wkb_geometry
    Left only
        Empty DataFrame
        Columns: [shape_area, wkb_geometry, lucats, shape_leng, project_na, ulurpno, effective, status]
        Index: []
    Right only
        Empty DataFrame
        Columns: [shape_area, wkb_geometry, lucats, shape_leng, project_na, ulurpno, effective, status]
        Index: []
```

## dof_dtm

Our first rounding errors!

I had to do a few things here. One, we had some crazy small rounding errors for shape_leng and shape_area. We don't use these columns at all. I rounded each to the nearest 8 decimal places manually, then ran comparison

I then had a bunch of "mismatches" due to duplicate keys. Then, I added `shape_leng` (to bbl, block, lot) as key. This got rid of duplicate keys, and the fact that we have no missing keys from either dataframe and no changes for any of the keys seems decisive to me.
```
________________________________________________________________________________
Tables
    Left: dof_dtm_library
    Right: dof_dtm_ingest
________________________________________________________________________________
Row count
    Left: 858158
    Right: 858158
________________________________________________________________________________
Column comparison
    Both
        air_lot_fl
        bbl
        bill_bbl_f
        block
        boro
        condo_flag
        conversion
        created_da
        created_us
        data_library_version
        easement_f
        effective_
        last_edi_1
        last_edite
        lot
        lot_note
        nycmap_bld
        ogc_fid
        reuc_flag
        shape_area
        shape_leng
        sub_lot_fl
        value_refl
        wkb_geometry
    Left only: None
    Right only: None
    Type differences
        Bbl
            Left: character varying
            Right: text
        Effective 
            Left: character varying
            Right: text
        Created us
            Left: character varying
            Right: text
        Last edite
            Left: character varying
            Right: text
        Air lot fl
            Left: character varying
            Right: text
        Sub lot fl
            Left: character varying
            Right: text
        Easement f
            Left: character varying
            Right: text
        Condo flag
            Left: character varying
            Right: text
        Boro
            Left: character varying
            Right: text
        Lot note
            Left: character varying
            Right: text
        Reuc flag
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Key columns
        bbl
        block
        lot
        shape_leng
    Left only: None
    Right only: None
    Duplicate keys: None
    Columns with diffs: None
    Differences by column: None
```

## dcp_zoningmapindex
```
________________________________________________________________________________
Tables
    Left: dcp_zoningmapindex_library
    Right: dcp_zoningmapindex_ingest
________________________________________________________________________________
Row count
    Left: 131
    Right: 131
________________________________________________________________________________
Column comparison
    Both
        data_library_version
        eastmap
        northeastm
        northmap
        northwestm
        ogc_fid
        quartersec
        section
        shape_area
        shape_leng
        southeastm
        southmap
        southwestm
        westmap
        wkb_geometry
        zoning_map
    Left only: None
    Right only: None
    Type differences
        Northmap
            Left: character varying
            Right: text
        Section
            Left: character varying
            Right: text
        Northeastm
            Left: character varying
            Right: text
        Northwestm
            Left: character varying
            Right: text
        Southwestm
            Left: character varying
            Right: text
        Westmap
            Left: character varying
            Right: text
        Quartersec
            Left: character varying
            Right: text
        Southeastm
            Left: character varying
            Right: text
        Southmap
            Left: character varying
            Right: text
        Zoning map
            Left: character varying
            Right: text
        Eastmap
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Compared columns
        eastmap
        northeastm
        northmap
        northwestm
        quartersec
        section
        shape_area
        shape_leng
        southeastm
        southmap
        southwestm
        westmap
        wkb_geometry
        zoning_map
    Left only
        Empty DataFrame
        Columns: [shape_area, northmap, section, shape_leng, northeastm, northwestm, southwestm, westmap, quartersec, wkb_geometry, southeastm, southmap, zoning_map, eastmap]
        Index: []
    Right only
        Empty DataFrame
        Columns: [shape_area, northmap, section, shape_leng, northeastm, northwestm, southwestm, westmap, quartersec, wkb_geometry, southeastm, southmap, zoning_map, eastmap]
        Index: []
```

## dcp_zoningtaxlots
```
________________________________________________________________________________
Tables
    Left: dcp_zoningtaxlots_library
    Right: dcp_zoningtaxlots_ingest
________________________________________________________________________________
Row count
    Left: 858088
    Right: 858088
________________________________________________________________________________
Column comparison
    Both
        bbl
        borough_code
        commercial_overlay_1
        commercial_overlay_2
        data_library_version
        limited_height_district
        ogc_fid
        special_district_1
        special_district_2
        special_district_3
        tax_block
        tax_lot
        zoning_district_1
        zoning_district_2
        zoning_district_3
        zoning_district_4
        zoning_map_code
        zoning_map_number
    Left only: None
    Right only: None
    Type differences
        Commercial overlay 2
            Left: character varying
            Right: text
        Borough code
            Left: character varying
            Right: text
        Limited height district
            Left: character varying
            Right: text
        Special district 3
            Left: character varying
            Right: text
        Tax lot
            Left: character varying
            Right: text
        Tax block
            Left: character varying
            Right: text
        Zoning district 4
            Left: character varying
            Right: text
        Special district 1
            Left: character varying
            Right: text
        Commercial overlay 1
            Left: character varying
            Right: text
        Zoning district 1
            Left: character varying
            Right: text
        Bbl
            Left: character varying
            Right: text
        Zoning district 2
            Left: character varying
            Right: text
        Zoning district 3
            Left: character varying
            Right: text
        Zoning map code
            Left: character varying
            Right: text
        Special district 2
            Left: character varying
            Right: text
        Zoning map number
            Left: character varying
            Right: text
________________________________________________________________________________
Data comparison
    Key columns
        bbl
        borough_code
        tax_block
        tax_lot
    Left only: None
    Right only: None
    Columns with diffs: None
    Differences by column: None
```

### PR DESCRIPTION
**Pull request by fvankrieken**
Closes #1246. Reports for specific datasets can be found there.

These are all datasets that I would expect to be more consistent in their schemas - I'm going to leave all the columns

See https://github.com/NYCPlanning/data-engineering/pull/1191/commits/180c7404762323113cff4a34d06fe22c2d5fa701 and the last two commits for testing for the only code changes - adding a new source type. There's a discussion to be had there, but per [this comment](https://github.com/NYCPlanning/data-engineering/pull/1191#discussion_r1851029300) I sort of think it's worth punting

### CODE REVIEW

File: dcpy/lifecycle/ingest/configure.py [hunk 1]
```diff
@@ -67,6 +68,13 @@ def get_version(source: Source, timestamp: datetime | None = None) -> str:
             return extract_socrata.get_version(source)
         case GisDataset():
             return publishing.get_latest_gis_dataset_version(source.name)
+        case DEPublished():
```
Comment (fvankrieken) [inline dcpy/lifecycle/ingest/configure.py:71]: We only have one of these for now. There's some discussion to be had for how we want to do this exactly - ideally, it's coming from `package` folders I think. But I think this is an issue worth punting - again, we currently only get ztl straight from our own output folders

**Comment by fvankrieken** [comment]
Reran all datasets through ingest and reran comparison after rebasing in latest main with my vaildation tweaks. Looks good to go! Check out #1246 for details on datasets, I think it's important that reviewers look through each in case the migrater missed anything.

As a norm for future PRs, should we have those reports in the PR, each as their own comment? Might make it easier to follow/track

**Comment by fvankrieken** [comment]
Oops - missed one dataset that comes from edm-publishing/datasets. Going to do that real quick

**Comment by fvankrieken** [comment]
Ready to go

File: dcpy/lifecycle/ingest/templates/dof_condo.yml
Comment (damonmcc) [inline dcpy/lifecycle/ingest/templates/dof_condo.yml:1]: can you rebase this into https://github.com/NYCPlanning/data-engineering/pull/1191/commits/b7073fd04a56490653e29cd3822a6f69a510fb19?

**Comment by damonmcc** [comment]
> As a norm for future PRs, should we have those reports in the PR, each as their own comment? Might make it easier to follow/track

each report as their own comment sounds great

hard to pick between putting them in the issue or the PR though. I think in the issue would be nice to keep the PR page small, but no strong feelings either way

File: dcpy/lifecycle/ingest/templates/dof_condo.yml
Comment (fvankrieken) [reply dcpy/lifecycle/ingest/templates/dof_condo.yml:1]: I was gonna squash the whole thing

**Review by damonmcc — APPROVED** [review]

**Pull request merged**